### PR TITLE
[ADD] web widget bootstrap toggle

### PR DESCRIPTION
**Pull request by aimsystems**
Added the web_toggle directory which when installed allows for the use of a bootstrap toggle widget within odoo's backend. It is meant to be used for booleans to replace checkboxes with more attractive Yes/No or On/Off toggle's and other options as well.

### CODE REVIEW

**Comment by petrus-v** [comment]
That's shame, I've done quite the same taff against v7 #236

**Comment by aimsystems** [comment]
Well I havent looked at your V7 but its great really if all of us can expect to find compatible widgets for all of the odoo versions. So if your V7 widget is adopted and my V8 (maybe 9) is accepted then we have covered all the major odoo versions.  I still need to take the time to determine how to overcome the Travis CI build failure which I know nothing about as of now.

File: web_toggle/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
+    'name' : 'AIM Bootstrap Toggle Widget',
+    'version': '1.0',
+    'summary': 'Odoo Checkbox Widget',
+    'category': 'Tools',
+    'description':
+        """
+AIM Checkbox Toggle Widget
```
Comment (legalsylvain) [inline web_toggle/__openerp__.py]: please remove AIM reference in name and description.

**Comment by legalsylvain** [comment]
@petrus-v  & @aimsystems : It should be great to name both modules with the same name, and to set the same type of call. (widget="boolean_switch" ...)
So, migration of modules using this module from 7.0 to 8.0 will not require extra work.

@petrus-v : read [ML](https://www.odoo.com/fr_FR/groups/community-59/community-13769111?mode=thread) !

**Comment by aimsystems** [comment]
I have changed the widget to be enabled using widget="boolean_switch"

File: web_toggle/__openerp__.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
+{
+    'name': 'Bootstrap Toggle Widget',
+    'version': '1.0',
```
Comment (dreispt) [inline web_toggle/__openerp__.py]: `8.0.1.0.0`

File: web_toggle/__openerp__.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    'summary': 'Odoo Checkbox Widget',
+    'category': 'Tools',
+    'description':
```
Comment (dreispt) [inline web_toggle/__openerp__.py]: Missing author and license keys

File: web_toggle/static/src/js/bootstrap-toggle.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
+ +function ($) {
```
Comment (dreispt) [inline web_toggle/static/src/js/bootstrap-toggle.js]: Does this have lib have a license file to attach to it?

File: web_toggle/__openerp__.py (hunk 3 above)
Comment (dreispt) [inline web_toggle/__openerp__.py]: Make the description a README.rst file (remember to use the OCA template)

File: web_toggle/static/src/js/bootstrap-toggle.js (hunk 4 above)
Comment (aimsystems) [inline web_toggle/static/src/js/bootstrap-toggle.js]: After reviewing the zip file that bootstrap gives you as a download I found a LICENCE file. I have added it next to the to the two files I use from the bootstrap-toggle library.

File: web_toggle/__openerp__.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+{
+    'name': 'Bootstrap Toggle Widget',
+    'version': '8.0.1.0.0',
+    'author': "aimsystems,Odoo Community Association (OCA)",
+    'license': 'AGPL-3',
+    'website': 'https://github.com/OCA/web',
+    'summary': 'Odoo Checkbox Toggle Widget',
+    'category': 'Tools',
+    'description':
```
Comment (dreispt) [inline web_toggle/__openerp__.py]: Can you remove the description from here and add an OCA README.rst including it's content?

Comment (aimsystems) [inline web_toggle/__openerp__.py]: I have removed the description. Waiting for the build to run its checks. Thanks for guiding me through this Daniel.

File: web_toggle/README.rst [hunk 6]
```diff
@@ -0,0 +1,66 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License
+
+Boolean Bootstrap Toggle
+===================
```
Comment (dreispt) [inline web_toggle/README.rst]: Fix this, or rst to html conversion might fail:

```
Boolean Bootstrap Toggle
========================
```

File: web_toggle/README.rst [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+       options="{
+                 'data-on':'Yes',
+                 'data-off':'No',
+                 'data-onstyle':'primary',
+                 'data-offstyle':'danger',
+                 'data-size':'mini'
+                }"/>
+
+The above example would create a mini size swich which displayed a 'Yes' when true and 'No' when false would 
+display bootstrap's primary class colour when true and the danger class colour when false.
```
Comment (dreispt) [inline web_toggle/README.rst]: Great!

**Comment by aimsystems** [comment]
I have updated my project. The tests seem to have all passed. I am not sure what the protocol is to initiate the next stage of acceptance. Any input would be appreciated.

**Comment by dreispt** [comment]
Next step is simply to ask for review again, as you did.

File: web_toggle/README.rst [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+Usage
+=====
+
+After you installed it, you can add the text widget='boolean_switch' to any boolean within a form view. This 
+will change the checkbox to an 'On' or 'Off' button which slide from one to the other upon clicking. There are 
+many options within the bootstrap-toggle library which you can use to make additional changes. They can all be 
+utilized by providing an options dictionary. Go to bootstrap-toggle documentation to learn about all available 
+options such as colours, size, text, shapes, etc. Here is a full example usage.
+
+<field name="my_boolean" widget='boolean_switch' options="{'data-on':'Yes','data-off':'No','data-onstyle':'primary','data-offstyle':'danger','data-size':'mini'}"/>
```
Comment (dreispt) [inline web_toggle/README.rst]: It would be nice to format this [as code](http://sphinx-doc.org/rest.html#source-code), but I won't pick on that.

**Comment by dreispt** [comment]
:+1:

**Comment by aimsystems** [comment]
Thanks a lot Daniel. I am new to this so I am not sure what the etiquette is regarding turn around time on project stages. I know that we all have a lot of work to do. I do not mind updating the formatting to represent code. I will do so tonight. Just finishing up a little something for work and I will get to it tonight. Thanks again!

**Comment by dreispt** [comment]
We're all volunteers, no mandatory times here. You return to it when you can, and reviewers look at it again when they can - if not the former ones, maybe new ones. It's not that rigid.

**Comment by aimsystems** [comment]
Have updated the README.rst to format the code example as such. Whenever someone has an opportunity to review again I would appreciate it. 

Thanks again!

**Comment by aimsystems** [comment]
Hi OCA, I have completed updating my branch. How do I have my widget added to the OCA/web project?

I apologize but this is my first contribution and I am learning as I go here. 

Thanks for your input!

**Comment by dreispt** [comment]
You already have my :+1:
Another review is needed.

**Comment by dreispt** [comment]
@aimsystems the thing is, anyone can do it not many people contribute with reviews.

**Comment by legalsylvain** [comment]
@aimsystems : thanks for your contribution. You can discover OCA coding rules [here](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md) and OCA sharing rules [here](https://odoo-community.org/page/code). specially the section "Get your code merged".

@petrus-v : could you take a look and give your PoV on this PR, as you worked on similar module in V7.

Kind regards.

File: web_toggle/README.rst [hunk 9]
```diff
@@ -0,0 +1,60 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License
+
```
Comment (pedrobaeza) [inline web_toggle/README.rst]: Please add here the target link for the badge (according latest template), an upper line for highlighting the title, and other new sections like Try me on runbot. You can see the full template here: https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst

File: web_toggle/static/src/css/LICENCE [hunk 10]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
Comment (pedrobaeza) [inline web_toggle/static/src/css/LICENCE]: Is MIT license compatible with AGPL? I think so, but I want to be sure

**Comment by pedrobaeza** [comment]
Thanks for the contribution. I hope this is the first of plenty ones!

I have reviewed the PR, and I would like you to improve these things:
- Change the name to _web_widget_bootstrap_toggle_, because all the widgets addons have this prefix in the name. Including also bootstrap word, you help to specify which king of toggle it is.
- You should mention in any place from which bootstrap version you have extracted the toggle button. Odoo already ships a bootstrap version. Isn't included in this bundle?
- You can provide a view as demo data that changes one existing view on base module that switch the widget for a boolean field. This way, it's not needed to modify nothing to try your module.

**Comment by aimsystems** [comment]
Thanks guys. I Will implement your suggestions over the weekend. Thanks again for the prompt feedback. 

Much appreciated.

File: web_toggle/static/src/js/bootstrap-toggle.js (hunk 4 above)
Comment (petrus-v) [inline web_toggle/static/src/js/bootstrap-toggle.js]: Hi!

Could you move third party library under /static/src/lib directory ?

File: web_toggle/static/src/js/bs_toggle.js [hunk 11]
```diff
@@ -0,0 +1,74 @@
+openerp.web_toggle = function(instance, local) {
+    var _t = instance.web._t,
+        _lt = instance.web._lt;
```
Comment (petrus-v) [inline web_toggle/static/src/js/bs_toggle.js]: I haven't seen where you are using those variables. If you don't need it, you can remove it;)

**Comment by aimsystems** [comment]
Yes, I will follow up on all suggestions over the weekend. 

Thanks again for you support!

File: web_toggle/static/src/js/bs_toggle.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    var QWeb = instance.web.qweb;
+
+local.FieldBooleanToggle = instance.web.form.AbstractField.extend({
+    template: 'FieldBooleanToggle',
+    start: function() {
+        var self = this;
+        this.$checkbox = $("input", this.$el);
```
Comment (petrus-v) [inline web_toggle/static/src/js/bs_toggle.js]: I didn't know that syntax `$("input", this.$el)`, I usually use `this.$el.find("input")`. I won't complain about I've tested it both works!

**Comment by petrus-v** [comment]
I've tested on runbot uning chromium (Version 46.0.2490.71 Built on 8.2, running on Debian 8.2 (64-bit)) on res.users form with this setting on active field:

```
<field name="active" widget="boolean_switch"/>
```

I got 3 issues
- In edit mode, when clicking the widget the value doesn't change
- When I leave edit mode discarding change, the widget is just a blue
  rectangle (loosing text and button)
- When I set a user inactive a change record using arrows on top of form view,
  the widget styled still showing the value from the first user.

Do not hesitate to have look what I've done in v7, this king of things haven't change a lot!

**Comment by aimsystems** [comment]
Hey guys I have been trying to implement the suggestions. Now my build keeps complaining about the fact that there is no module named web_widget_bootstrap_toggle. Would anyone be able to take a peak and see what it is that I need to change. Thanks

**Comment by aimsystems** [comment]
Also thanks petrus-v I have now implemented an example for the active field in users. Also found why it was not working. In my test application I was adding the boostrap-toggle.css and bootstrap-toggle.js which I have now added to the backend.

**Comment by pedrobaeza** [comment]
You have forgotten to include file **init**.py in the new module layout. Please also take the occasion to squash all the commits to one, as this is the PR that adds the module.

**Comment by aimsystems** [comment]
Thanks, how do I squash the commits?

**Comment by pedrobaeza** [comment]
You can do it with `git rebase -i --autosquash efb28067065b0a374422d1f68c13f080fc0203ad`

**Comment by aimsystems** [comment]
git rebase -i --autosquash efb28067065b0a374422d1f68c13f080fc0203ad
error: could not apply 2a5ec4e... Update toggle_widget.xml

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
Could not apply 2a5ec4e716b5005d2465754fea89c9979de8f679... Update toggle_widget.xml

**Comment by aimsystems** [comment]
Git is complaining. Im a real newb with git

**Comment by pedrobaeza** [comment]
I see. You have a total screwed branch. It's better to copy your module directory, clone again main branch, insert your module in one commit, and then upload it to this branch with a force push (`git push -f`)

**Comment by aimsystems** [comment]
Thanks. I need to take some time to review the documentation on proper GIT protocol and the OCA standards and best practices. I really just tried to jump into it. Which may have been a little overzealous.

**Comment by pedrobaeza** [comment]
I know it can be hard at the beginning. If you don't get it, just update this PR with latest changes, and I'll fix it up when merging.

**Comment by dreispt** [comment]
@aimsystems If you didn't `git push`, the server stored branch is intact. Assuming that `origin` points at your fork you can `git reset --hard origin/8.0`.

**Comment by aimsystems** [comment]
Thanks Daniel and Pedro.I will look through the documentation and get a better handle on GIT and the OCA standards.  Pedro, like you said I completely butchered my branch.

**Comment by pedrobaeza** [comment]
@aimsystems, any news on this? Are you going to finish the work?

**Comment by aimsystems** [comment]
Hi Pedro, you are right. I probably should see it through. I put it on 
the back burner as I had a lot on my plate. I will take a look at it 
next week and see what needs to be done.

**Pull request closed without merging**